pkg/koordlet/runtimehooks/hooks/coresched/core_sched.go

```diff
@@ -0,0 +1,468 @@
+		p.SetKubeQOSCPUIdle, reconciler.NoneFilter())
+	p.Setup(op)
+}
+
+func (p *Plugin) Setup(op hooks.Options) {
+	p.reader = op.Reader
+	p.executor = op.Executor
+	p.cse = sysutil.NewCoreSchedExtended()
+}
+
+func (p *Plugin) SystemSupported() (bool, string) {
+	if p.sysSupported == nil {
+		isSupported, msg := sysutil.IsCoreSchedSupported()
+		p.sysSupported = pointer.Bool(isSupported)
+		p.supportedMsg = msg
+		klog.Infof("update system supported info for plugin %s, supported %v, msg %s",
+			name, *p.sysSupported, p.supportedMsg)
+	}
+	return *p.sysSupported, p.supportedMsg
+}
+
+func (p *Plugin) InitCache(podMetas []*statesinformer.PodMeta) bool {
+	if p.initialized.Load() {
+		return true
+	}
+
+	synced := p.LoadAllCookies(podMetas)
+
+	p.initialized.Store(synced)
+	return synced
+}
+
+func (p *Plugin) IsCacheInited() bool {
+	return p.initialized.Load()
+}
+
+func (p *Plugin) SetKubeQOSCPUIdle(proto protocol.HooksProtocol) error {
+	kubeQOSCtx := proto.(*protocol.KubeQOSContext)
+	if kubeQOSCtx == nil {
+		return fmt.Errorf("kubeQOS protocol is nil for plugin %s", name)
+	}
+	kubeQOS := kubeQOSCtx.Request.KubeQOSClass
+
+	if !p.rule.IsInited() {
+		klog.V(5).Infof("plugin %s has not been inited, rule inited %v, aborted to set cpu idle for QoS %s",
+			name, p.rule.IsInited(), kubeQOS)
+		return nil
+	}
+
+	isCPUIdle := p.rule.IsKubeQOSCPUIdle(kubeQOS)
+	if isCPUIdle {
+		kubeQOSCtx.Response.Resources.CPUIdle = pointer.Int64(1)
+	} else {
+		kubeQOSCtx.Response.Resources.CPUIdle = pointer.Int64(0)
+	}
+
+	return nil
+}
+
+// SetContainerCookie reconciles the core sched cookie for the container.
+// There are the following operations about the cookies:
+//  1. Get: Get the cookie for a core sched group, firstly try to find in cache and then get from the existing PIDs.
+//  2. Add: Add a new cookie for a core sched group for a container, and add a new entry into the cache, ref count = 1.
+//  3. Assign: Assign a cookie of an existing core sched group for a container, and increase cookie's ref count in cache.
+//     The cookies of sibling PGIDs (i.e. the PGIDs of the same core sched group) will be fetched in the Assign. If all
+//     the cookies of sibling PGIDs are default or invalid, the Assign will fall back to Add.
+//  4. Clear: Clear a cookie of an existing core sched group for a container (reset to default cookie 0), and the
+//     cookie's reference count in the cache is decreased. The cache entry will be removed when the ref count is no
+//     larger than zero.
+//
+// If multiple non-default cookies are assigned to existing containers of the same group, the firstly-created and
+// available cookie will be retained and the PGIDs of others will be moved to the former.
+// NOTE: The agent itself should be set the default cookie. It can be excluded by setting QoS to SYSTEM.
+func (p *Plugin) SetContainerCookie(proto protocol.HooksProtocol) error {
```

Comment:
- add rateLimiter for safety
- tradeoff between /proc read and syscall

Reply:
no longer read /proc, switch to prctl for thread groups